src/Language/PureScript/TypeChecker.hs

```diff
@@ -155,18 +155,28 @@ checkDuplicateTypeArguments args = for_ firstDup $ \dup ->
   firstDup = listToMaybe $ args \\ nub args
 
 checkTypeClassInstance
-  :: (MonadState CheckState m, MonadError MultipleErrors m, MonadWriter MultipleErrors m)
-  => ModuleName
+  :: (MonadState CheckState m, MonadError MultipleErrors m)
+  => TypeClassData
+  -> Int -- type argument index
   -> Type
   -> m ()
-checkTypeClassInstance _ (TypeVar _) = return ()
-checkTypeClassInstance _ (TypeLevelString _) = return ()
-checkTypeClassInstance _ (TypeConstructor ctor) = do
-  env <- getEnv
-  when (ctor `M.member` typeSynonyms env) . throwError . errorMessage $ TypeSynonymInstance
-  return ()
-checkTypeClassInstance m (TypeApp t1 t2) = checkTypeClassInstance m t1 >> checkTypeClassInstance m t2
-checkTypeClassInstance _ ty = throwError . errorMessage $ InvalidInstanceHead ty
+checkTypeClassInstance cls i = check where
+  -- is this argument fully determined via fundeps
+  isFunDepDetermined = (Any False, Any True) == foldMap determining (typeClassDependencies cls)
+    where determining fd = (Any (i `elem` fdDeterminers fd),
+                            Any (i `elem` fdDetermined fd))
```

Comment:
Well let's write out explicitly what we want.

We want to prevent the user from casing on types in rows (i.e. labels in rows shouldn't determine behavior).

So we can't have two _different_ instances with different rows in some type argument and every other type argument the same. 

One way to achieve that with a single functional dependency (e.g. `MonadState`, where `m -> s`) is to say that we have a functional relationship on type arguments (`m` determines `s`), so changing the result of the function forces the argument to change (if `s` differs, `m` must differ too). If the input types are the same but the outputs differ, that will lead to overlap.

In general, there must be some type argument which is forced to differ between the two instances. We can find the set of such type arguments by following functional dependencies backwards, but I think to do this properly we need a full (static) overlap check, which is why I suggest limiting things to a single functional dependency for now.